doc/contributing/maintaining-shared-library-support.md

```diff
@@ -0,0 +1,125 @@
+
+On non-Windows platoforms, Node.js is built with the shared library
+option by adding `--shared` to the configure step. On Windows
+platofrms Node.js is built with the shared library option by
+adding `dll` to the vcbuild command line.
+
+Once built there are two key components:
+
+* executable - node
+* library - libnode
+
+The node executable is a thin wrapper around libnode which is
+generated so that we can run the standard Node.js test suite
+against the shared library.
+
+The executable and library will have extensions as appropriate
+for the platform on which they are built. For
+example, node.exe on windows and node on other platforms for
+the executable.
+
+libnode may have additional naming components, as an example
+in a build on macOS `libnode.105.dylib`.
+
+In cases where an application links against the shared
+library it is up to the application developer to add options
+so that the shared library can be found by the application or
+to set the LIBPATH (AIX), LD\_LIBRARY\_PATH (Linux/Unix), etc.
+so that it is found at runtime.
+
+For the node.exe wrapper, it is built so that it can
+find the shared library in one of the following:
+
+* the same directory as the node executable
+* ../lib with the expectation that the executable is
+  installed in a `bin` directory at the same level
+  as a `lib` directory in which the shared library is
+  installed. This is where the default package that
+  is build with the shared library option will
+  place the executable and library.
+
+## Exports
+
+On windows, functions that may be linked from native
+addons or additional Node.js executables need to have
+NODE_EXTERN_PRIVATE or NODE_EXTERN otherwise they will
+not be exported by the shared library. In the case of
+functions used by additional Node.js executables
+(ex: `mksnapshot`) a missing NODE_EXTERN or 
+NODE_EXTERN_PRIVATE will cause the build to fail.
+NODE_EXTERN_PRIVATE should be used in these cases
+unless the intent is to add the function to the
+public embedder API.
+
+## Native addons
+
+For regular Node.js builds, native addons rely on symbols
+exported by the node executable. As a result any
+pre-built binaries will expect symbols from the executable
+instead of the shared library itself.
+
+The current node executable wrapper addresses this by
+ensuring that node.lib from node.exe does not get generated
+and instead copy libnode.lib to node.lib. This means addons
+compiled when referencing the correct node.lib file will correctly
+depend on libnode.dll. The down side is that native addons compiled
+with stock Node.js will still try to resolve symbols against
+node.exe rather than libnode.dll. This would be a problem for
+package that use node-addon-api pre-compiled binaries with the
+goal of them working across different Node.js versions and
+types (standard versus shared library). At this point the
+main purpose for the node executable wrapper is for testing
+so this is not considered a major issue.
```

Comment:
@lux01 I assume we should also update this section as well since the
approach has been updated.